steps/expression_processor.go

```diff
@@ -91,26 +136,34 @@ func (p *ExpressionProcessor) String() string {
 	return res + ": " + str.String()
 }
 
-func (p *ExpressionProcessor) evaluate(sample *bitflow.Sample, header *bitflow.Header) (bool, error) {
+func (p *ExpressionProcessor) evaluate(sample *bitflow.Sample, header *bitflow.Header) (*bitflow.Sample, *bitflow.Header, error) {
 	if p.checker.HeaderChanged(header) {
 		for _, expr := range p.expressions {
 			if err := expr.UpdateHeader(header); err != nil {
-				return false, err
+				return nil, nil, err
 			}
 		}
 	}
+	outSample := sample
+	outHeader := header
+	var err error
+	var res bool
 	for _, expr := range p.expressions {
 		if p.Filter {
-			if res, err := expr.EvaluateBool(sample, header); err != nil {
-				return false, err
-			} else if !res {
-				return false, nil
+			res, err = expr.EvaluateBool(outSample, outHeader)
+			if err != nil {
+				return nil, nil, err
+			}
+			if ! res {
+				return nil, nil, nil
 			}
 		} else {
-			if _, err := expr.Evaluate(sample, header); err != nil {
-				return false, err
+			outSample, outHeader, err = expr.Evaluate(outSample, outHeader)
+			if err != nil {
```

Comment:
This error check is not necessary, you can simply return the result of the `Evaluate` method. This would also allow you to remove the four variable declarations (`outSample`, `outHeader`, `res` and `err`) by turning `=` assignments into `:=` assignments in various places.

Reply:
Still, the declaration are necessary imo. The expression evaluation is surrounded by a loop. I assume that successor expression evaluations must get the sample and header results of their predecessor. The final return value of this is the result of the last expression evaluation. Thus, the variables must be declared at the scope above the loop.

Reply:
Thanks for the explanation! I missed the loop.